py/server.py

```diff
@@ -145,21 +147,50 @@ def __init__(self):
         super(Application, self).__init__(handlers, **tornado_settings)
 
 
-def broadcast_envs(handler):
-    for s in handler.subs:
-        handler.subs[s].write_message(json.dumps(
+def broadcast_envs(handler, target_subs=None):
+    if target_subs is None:
+        target_subs = handler.subs.values()
+    for sub in target_subs:
+        sub.write_message(json.dumps(
             {'command': 'env_update', 'data': list(handler.state.keys())}
         ))
 
 
 class SocketHandler(tornado.websocket.WebSocketHandler):
     def initialize(self, state, subs):
+        self.layouts = self.load_layouts()
         self.state = state
         self.subs = subs
+        self.broadcast_layouts()
 
     def check_origin(self, origin):
         return True
 
+    def broadcast_layouts(self, target_subs=None):
+        if target_subs is None:
+            target_subs = self.subs.values()
+        for sub in target_subs:
+            sub.write_message(json.dumps(
+                {'command': 'layout_update', 'data': self.layouts}
+            ))
+
+    def save_layouts(self):
+        layout_filepath = os.path.join(FLAGS.env_path, 'view', LAYOUT_FILE)
+        fn = open(layout_filepath, 'w')
```

Comment:
Check that `fn` is actually open? (You could use the `with open(layout_filepath, 'w') as fn:` syntax here.)